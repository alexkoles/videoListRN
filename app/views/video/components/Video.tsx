import * as React from 'react';
import RNVideo from 'react-native-video';
import {StyleSheet, View} from 'react-native';
import {Colors} from 'react-native-paper';
import {
  TapGestureHandler,
  TapGestureHandlerStateChangeEvent,
  State,
} from 'react-native-gesture-handler';

interface VideoProps {
  uri: string;
  preview: string;
  onDoubleTap?: (point: { x: number; y: number }) => void;
  paused?: boolean;
  onPausedChanged?: (paused: boolean) => void;
}

export const Video = React.memo((props: VideoProps) => {
  const {
    uri = 'https://res.cloudinary.com/dl66ay16a/video/upload/v1599461416/videos/fox_qgjt9u.mp4',
    preview = 'https://res.cloudinary.com/dl66ay16a/image/upload/v1599230016/videos/Fox_je7rdx.png',
    onDoubleTap,
    onPausedChanged,
  } = props;
  const [paused, setPaused] = React.useState(props.paused);

  React.useEffect(() => {
    setPaused(props.paused);
  }, [props.paused]);
  const doubleTap = React.useRef(React.createRef<TapGestureHandler>().current);

  const onSingleTapped = (event: TapGestureHandlerStateChangeEvent) => {
    if (event.nativeEvent.state === State.ACTIVE) {
      const nextState = !paused;
      setPaused(nextState);
      onPausedChanged && onPausedChanged(nextState);
    }
  };

  const onDoubleTapped = (event: TapGestureHandlerStateChangeEvent) => {
    if (event.nativeEvent.state === State.ACTIVE) {
      onDoubleTap &&
        onDoubleTap({x: event.nativeEvent.x, y: event.nativeEvent.y});
    }
  };

  return (
    <TapGestureHandler
      waitFor={doubleTap}
      onHandlerStateChange={onSingleTapped}>
      <TapGestureHandler
        ref={doubleTap}
        onHandlerStateChange={onDoubleTapped}
        numberOfTaps={2}>
        <View style={styles.backgroundVideo}>
          <RNVideo
            repeat
            paused={paused}
            source={{uri}}
            poster={preview}
            posterResizeMode="cover"
            style={styles.video}
            resizeMode="cover"
          />
        </View>
      </TapGestureHandler>
    </TapGestureHandler>
  );
});

const styles = StyleSheet.create({
  backgroundVideo: {
    flex: 1,
    backgroundColor: '#000000',
  },
  video: {
    flex: 1,
  },
});
