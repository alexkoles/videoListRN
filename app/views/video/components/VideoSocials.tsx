import * as React from 'react';
import {StyleSheet, View} from 'react-native';
import {iVideoStream, iPointStamp, iPoint} from 'app/utils/dataType';
import {Video} from './Video';
import {Comment} from './Comment';
import {HeartAnimationFullView} from './heartAnimation';
import {Colors} from 'react-native-paper';

interface VideoSocialsProps {
  data: iVideoStream;
  paused: boolean;
  onPausedChanged: (paused: boolean) => void;
}

export const VideoSocials: React.FC<VideoSocialsProps> = (props) => {
  const [points, setPoints] = React.useState<iPointStamp[]>([]);
  const {data, paused, onPausedChanged} = props;
  // console.log('HERE_DATA:' + JSON.stringify(data));

  const onVideoDoubleTap = React.useCallback((point: iPoint) => {
    setPoints((pre) =>
      pre.concat(Object.assign({timestamp: Date.now()}, point)),
    );
  }, []);

  return (
    <View style={styles.container}>
      <Video
        uri={data.uri}
        preview={data.preview}
        paused={paused}
        onPausedChanged={onPausedChanged}
        onDoubleTap={onVideoDoubleTap}
      />
      <Comment videoId={data.id} comment={data.comment} style={styles.social} />
      <HeartAnimationFullView style={styles.heart} points={points} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.yellow900,
  },
  social: {
    position: 'absolute',
    right: 16,
    bottom: 16,
  },
  heart: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
  },
});
