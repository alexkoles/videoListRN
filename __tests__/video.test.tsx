import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {Video} from 'app/views/video/components/Video';

jest.mock('react-native-video', () => 'Video');

describe('Video Component', () => {
  const url =
    'https://res.cloudinary.com/dl66ay16a/video/upload/v1599461415/videos/dogs_a24lmq.mp4';
  const preview =
    'https://res.cloudinary.com/dl66ay16a/image/upload/v1599229049/videos/dogs_qpukey.png';
  it('renders video correctly', () => {
    const component = renderer
      .create(<Video uri={url} preview={preview} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
