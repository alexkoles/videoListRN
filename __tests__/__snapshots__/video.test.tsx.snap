// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video Component renders video correctly 1`] = `
<View
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <Video
    poster="https://res.cloudinary.com/dl66ay16a/image/upload/v1599229049/videos/dogs_qpukey.png"
    posterResizeMode="cover"
    repeat={true}
    resizeMode="cover"
    source={
      Object {
        "uri": "https://res.cloudinary.com/dl66ay16a/video/upload/v1599461415/videos/dogs_a24lmq.mp4",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
